keps/sig-api-machinery/2019-02-13-custom-resource-definitions.md

```diff
@@ -0,0 +1,212 @@
+Currently there is no good way to express that in order for CustomResource `A` to work correctly, 
+CustomResource `B` needs to be installed as well.
+
+#### Version upgrades
+
+Another open question for CustomResources is version management. Currently there are no built-in practices
+for moving from one version of a CustomResource to another. To do such a migration, the tool itself would
+have to perform the schema migration by reading all of the data objects in the old version and rewriting
+them in the new version. To make this successful , at the very least we will need to provide the ability to
+mark a CustomResource as "read-only" to ensure that there isn't skew between the two representations.
+
+We may need to invest significantly more heavily in this space to make it easy for users to upgrade or 
+update their objects. For now, most users have simply chosen not to upgrade the version.
+
+### Risks and Mitigations
+
+#### DDOS or other attacks on storage
+
+Because custom resources share the same storage as the core API types, there is risk that either via 
+requests/second or storage space usage, custom resources will become deliberate or inadvertent denial
+of service attacks on a cluster. Currently there is a maximum size cap for custom resources, but it is
+a pretty limited defense against these sorts of concerns. This makes the installation of a custom
+resource a very privileged action for a cluster.
+
+#### Custom resource adminstration
+
+For many reasons, installing a new custom resource is a privileged action for a cluster adminstrator.
+However, this significantly increases the complexity of developing a Custom Resource. Ideally the
+resource itself would install itself into the cluster when it's controller is run as a Pod, but
+because you don't want to grant the controller the general ability to create arbitrary CustomResources,
+it is generally the case that the controller won't have proper permissions to create a resource. 
+
+In the future we may want to develop some sort of "single use token" or some equivalent such that custom resources are easy to install, and yet don't require cluster adminstrator privileges to bootstrap themselves.
+
+## Graduation Criteria
+
+This work has already graduated.
+
+## Drawbacks
+
+To become full-fledged API objects CustomResources require extensive usage of webhooks for various API 
+object lifecycle events (e.g. validation, defaulting, etc). This pattern means that the code for an custom 
+resource can be splattered around in various callbacks and as such can be harder to understand.
+
+Likewise, all types of APIs may not be implementable (or easily implementable) as custom resources because
+of this limitation.
+
+## Alternatives
+
+Aggregated API Servers where requests are delegated from the API server to an delegate API server
+are a more flexible (though more complicated) alternative to custom resource definitions.
```

Comment:
They are more complex for users to make use of, but they are actually a much simpler mechanism--we took them to GA without difficulty, CRDs have been much more work.

Reply:
@lavalamp so, they are more complicated for users, right? I wonder what this would all look like if API Servers were easier for users to use?

Reply:
If we made apiservers super easy to produce, we would likely have a major compatibility problem at upgrade time. CRDs are lower maintenance for users. I think this is a fairly fundamental difference: apiservers run a full stack. CRDs only need to change if you evolve your API; apiservers need to change with the Kubernetes version.

Reply:
(As Brian would say, "Aggregated apiservers are for *us*. We need them to build Kubernetes.")

Reply:
API aggregation was used to implement CRDs. The original use case was an API with a non-etcd backend (core metrics API). API aggregation, which reached GA in 1.10, was also simpler to achieve full functionality.

Reply:
Separate etcd storage is an even more critical concern for apiservers than it is for CRs, might be worth mentioning it as a risk.